### PR TITLE
feat(stepfunctions): add support for Map state

### PR DESCRIPTION
**Pull request by melalawi**
Today, Step Functions launched Map states. This PR adds support for this new state type in CDK.

https://aws.amazon.com/about-aws/whats-new/2019/09/aws-step-functions-adds-support-for-dynamic-parallelism-in-workflows/

https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-map-state.html

:)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by rix0rrr** [comment]
Duplicate: https://github.com/aws/aws-cdk/pull/4145

Will need some time to go over both and compare.

File: packages/@aws-cdk/aws-stepfunctions/lib/states/map.ts [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+     * @default $
+     */
+    readonly itemsPath?: string;
+
+    /**
+     * The JSON that you want to override your default iteration input
+     *
+     * @default $
+     */
+    readonly parameters?: { [key: string]: any };
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-stepfunctions/lib/states/map.ts]: Don't see this parameter documented here: https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-map-state.html

Are you sure it exists?

File: packages/@aws-cdk/aws-stepfunctions/lib/states/map.ts [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+    /**
+     * MaxConcurrency
+     *
+     * An upper bound on the number of iterations you want running at once. Default value of 0
+     *
+     * @default false
+     */
+    readonly sequential?: boolean;
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-stepfunctions/lib/states/map.ts]: Rather have the number here, no?

File: packages/@aws-cdk/aws-stepfunctions/lib/states/map.ts (hunk 1 above)
Comment (melalawi) [reply packages/@aws-cdk/aws-stepfunctions/lib/states/map.ts]: It does! You can use the [context object](https://docs.aws.amazon.com/step-functions/latest/dg/input-output-contextobject.html#contextobject-map) in map state. I'll talk to our technical writer to see if we can clarify the docs here.

File: packages/@aws-cdk/aws-stepfunctions/lib/states/map.ts (hunk 2 above)
Comment (melalawi) [reply packages/@aws-cdk/aws-stepfunctions/lib/states/map.ts]: Good catch! This is a remnant of a beta version of map state! :)

I'll update it to an unsigned integer.

**Comment by rix0rrr** [comment]
We need someone to carry this to the finish line :)

**Comment by rhboyd** [comment]
Is the other PR not being worked on? I can finish this up this weekend if the other PR has been abandoned.

**Comment by yyolk** [comment]
Would I be able to use this branch as-is to generate the subsequent python dependency with the docker image?

Right now my stack is in python and I've only tried building it once (it worked, Jsii is pretty neat)

I've only been installing from pip& now, so I'll at the very least be building this to see if it works for my needs - I can also report back if that helps get this merged?

**Comment by yyolk** [comment]
w00t!

**Pull request merged**